### PR TITLE
Insert accounts

### PR DESCRIPTION
**Pull request by carloszanella**
Some notes:
- I did some interface changes that felt required. you can see them on storage/mod.rs
- I will add some more assertions and tests in the next commits
- This is just an insert, I can try doing an upsert next.. not sure it will be needed?

### CODE REVIEW

File: src/storage/postgres/contract_state.rs [hunk 1]
```diff
@@ -91,8 +91,97 @@ where
...
+                    .first::<(i64, NaiveDateTime)>(db)
+                    .await
+                    .map_err(|err| {
+                        StorageError::from_diesel(err, "Transaction", &tx_hash.to_string(), None)
+                    })?;
+                created_ts = Some(_created_ts);
+                Some(tx_id)
+            }
+            None => None,
+        };
```
Comment (carloszanella) [inline src/storage/postgres/contract_state.rs]: this feels a bit hacky but Idk how to like get `tx_hash` and `created_ts` only if there is a creation transaction

File: src/storage/postgres/contract_state.rs [hunk 2]
```diff
@@ -91,8 +91,97 @@ where
...
+                .map_err(|err| {
+                    StorageError::from_diesel(err, "ContractCode", &new.address.to_string(), None)
+                })?;
+
+            Self::upsert_slots(
+                self,
+                ContractId(new.chain, new.address.as_bytes().to_vec()),
+                tx_hash.as_bytes(),
+                &new.slots,
+            )
```
Comment (carloszanella) [inline src/storage/postgres/contract_state.rs:181]: this method right now is not implemented

File: src/storage/mod.rs [hunk 3]
```diff
@@ -480,7 +480,11 @@ pub trait ContractStateGateway {
     ///
     /// # Parameters
     /// - `new` the new contract state to be saved.
-    async fn add_contract(&mut self, new: Self::ContractState) -> Result<(), StorageError>;
+    async fn add_contract(
+        &self,
+        new: &Self::ContractState,
+        db: &mut Self::DB,
+    ) -> Result<i64, StorageError>;
```
Comment (carloszanella) [inline src/storage/mod.rs:487]: Idk why self was `mut` here and I made `&new` a reference. Not sure if you had a special reason for it to be moved

File: src/storage/postgres/contract_state.rs [hunk 4]
```diff
@@ -30,77 +30,158 @@ where
...
+        let now = chrono::Utc::now().naive_utc();
+        let chain_id = self.get_chain_id(new.chain);
+        let mut tx_hash = H256::zero();
+        let mut created_ts = None;
+        let creation_tx_id = match new.creation_tx {
+            Some(tx) => {
+                tx_hash = tx;
+                let (tx_id, _created_ts) = schema::transaction::table
+                    .filter(schema::transaction::hash.eq(tx_hash.as_bytes()))
+                    .select((schema::transaction::id, schema::transaction::inserted_ts))
```
Comment (kayibal) [inline src/storage/postgres/contract_state.rs]: This should be the `block.ts`, not the time at which we inserted the transaction. We have a few options:
1. We retrieve the block through txhash like you tried here
2. We save the block or the timestamp as part of the model
3. We pass created_at: Block through the interface, but for upserts that would be strange

File: src/storage/postgres/contract_state.rs [hunk 5]
```diff
@@ -30,77 +30,158 @@ where
...
+                    .first::<(i64, NaiveDateTime)>(db)
+                    .await
+                    .map_err(|err| {
+                        StorageError::from_diesel(err, "Transaction", &tx_hash.to_string(), None)
+                    })?;
+                created_ts = Some(_created_ts);
+                Some(tx_id)
```
Comment (kayibal) [inline src/storage/postgres/contract_state.rs]: This is a bit weird, probably better to return a tuple. But indeed a weird case..

File: src/storage/postgres/contract_state.rs [hunk 6]
```diff
@@ -30,77 +30,158 @@ where
...
+            })?;
+
+        // Insert initial account balances to the respective table.
+        let mut balance_bytes = [0; 32];
+        new.balance.to_big_endian(&mut balance_bytes);
+
+        let orm_balance = orm::NewAccountBalance {
+            account_id: acc_id,
+            balance: balance_bytes.to_vec(),
+            valid_from: now,
```
Comment (kayibal) [inline src/storage/postgres/contract_state.rs]: This needs to correspond to the `block.ts` of the insertion transaction. Remember we have to insert historical data so it can't be just `now` here - we might be inserting a contract that was created years ago.

File: src/storage/postgres/contract_state.rs (hunk 5 above)
Comment (kayibal) [reply src/storage/postgres/contract_state.rs]: Maybe this?
```rust
let (creation_tx_id, created_ts): (Option<i64>, Option<NaiveDateTime>) = 
    match new.creation_tx {
        Some(tx) => {
            let (tx_id, _created_ts) = schema::transaction::table
                    .filter(schema::transaction::hash.eq(tx.as_bytes()))
                    .select((schema::transaction::id, schema::transaction::inserted_ts))
                    .first::<(i64, NaiveDateTime)>(db)
                    .await
                    .map_err(|err| {
                        StorageError::from_diesel(err, "Transaction", &tx.to_string(), None)
                    })?;
            (Some(tx_id), Some(_created_ts))
        }
        None => (None, None),
    };
```

File: src/storage/postgres/contract_state.rs (hunk 4 above)
Comment (kayibal) [reply src/storage/postgres/contract_state.rs]: We will get all data block & tx scoped from streamingfast - so it will look sth like this:
```
{
    "block": {"number": ..., "hash": ...},
    "transactions": [{
         "hash": "0x...",
         "created_contracts": [{...}, ..],
         "code_changes": [..],
         "balance_changes": [..],
         "storage_changes": [..],
    }],
}
```

File: src/storage/postgres/mod.rs [hunk 7]
```diff
@@ -332,22 +332,30 @@ mod fixtures {
...
+            data.push((
+                block_id.eq(b),
+                index.eq(i),
+                hash.eq(H256::from_str(h)
+                    .expect("valid txhash")
+                    .as_bytes()
+                    .to_owned()),
+                from.eq(from_val.as_bytes()),
+                to.eq(to_val.as_bytes()),
+                inserted_ts.eq(block.ts),
```
Comment (kayibal) [inline src/storage/postgres/mod.rs]: Is there any logic on this field? Or why did you add it? The transaction timestamp is the same as the block assigned to it. Both `inserted_ts` and `modified_ts` are maintained by Postgres triggers, so no need to set it them.

File: src/storage/postgres/contract_state.rs [hunk 8]
```diff
@@ -98,23 +98,24 @@ where
...
                     .map_err(|err| {
-                        StorageError::from_diesel(err, "Transaction", &tx_hash.to_string(), None)
+                        StorageError::from_diesel(err, "Transaction", &tx.to_string(), None)
                     })?;
-                created_ts = Some(_created_ts);
-                Some(tx_id)
+                (Some(tx_id), Some(_created_ts))
             }
-            None => None,
+            None => (None, Some(now)),
```
Comment (kayibal) [inline src/storage/postgres/contract_state.rs]: Does it make sense to wrap created_ts with Some() it seems you unwrap it most of the times below, doing it the other way around would probably be nicer. Retrieve without Option and wrap in Some if needed.
